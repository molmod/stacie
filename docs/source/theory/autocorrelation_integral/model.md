# Model Spectrum

Stacie supports two models for fitting the low-frequency part of the power spectrum.
In both models, the value at zero frequency corresponds to the autocorrelation integral.

1. The [ExpPolyModel](#stacie.model.ExpPolyModel) is the most general:
   it is an exponential function of a linear combination of simple monomials of the frequency.
   One can specify the degrees of the monomials and typically a low degree works fine:

    - Degree `[0]` is suitable for a white noise spectrum.
    - Degrees `[0, 1]` can be used to extract useful information of a very noisy spectrum.
    - Degrees `[0, 1, 2]` is applicable to spectra with low statistical uncertainty,
      e.g. averaged over 100 inputs.
    - An even polynomial with degrees `[0, 2]` is suitable for spectra with a peak at zero frequency.

    The main advantage of this model is its broad applicability,
    as it requires little prior knowledge of the functional form of the spectra.

2. The Exponential Tail model is designed for autocorrelation functions that decay exponentially.
   Its primary advantage is that, in addition to the integrated correlation time,
   it also provides an estimate of the exponential correlation time.

## 1. ExpPolyModel

The [ExpPolyModel](#stacie.model.ExpPolyModel) is defined as

$$
    C^\text{exppoly}_k = \exp\left(\sum_{s\in S} a_s f_k^s\right)
$$

where $S$ is the set of polynomial degrees.
With this form, $a_0$ corresponds to the integral of the autocorrelation function.
When one obtains an estimate $\hat{a}_0$ and an estimated variance $\hat{C}(a_0)$,
the autocorrelation integral is [log-normally distributed](https://en.wikipedia.org/wiki/Log-normal_distribution)
with estimated mean and variance:

$$
    \hat{\mathcal{I}}
    &= \exp\left(\hat{a}_0 + \frac{1}{2}\hat{C}(a_0)\right)
    \\
    \hat{C}(\mathcal{I})
    &= \exp\left(2\hat{a}_0 + \hat{C}(a_0)\right)
    \left(\exp(\hat{C}(a_0)) - 1 \right)
$$

## 2. Exponential Tail Model

This model represents the autocorrelation function
as the sum of a short-term component and a (periodic) exponentially decaying tail:

$$
c_\Delta = c_\Delta^\text{short} + c_\Delta^\text{tail}
$$

### Short-term component

The short-term component is non-zero only for small positive or negative time lags:

$$
c_\Delta^\text{short} \neq 0 & \quad\text{if } |\Delta| \le \Delta_\text{short}
\\
c_\Delta^\text{short} = 0 & \quad\text{if } \Delta_\text{short} \lt |\Delta| \le N/2
$$

Because summations in DFT are commonly taken from $0$ to $N-1$, we rewrite this as:

$$
c_\Delta^\text{short}
    \neq 0 &\qquad \forall\,\Delta\in\lbrace
        0, \ldots, \Delta_\text{short},
        N-\Delta_\text{short}, \ldots, N-1
    \rbrace
\\
c_\Delta^\text{short}
    = 0 &\qquad \forall\,\Delta\in\lbrace
        \Delta_\text{short}+1, \ldots, N-\Delta_\text{short} -1
    \rbrace
$$

### Tail Component

The tail component is the periodic repetition (sum over $i$) of two exponential functions:
one decaying for positive time lags and one for negative time lags.
For $\Delta \in \lbrace 0, \ldots, N-1 \rbrace$, this can be written as:

$$
c_\Delta^\text{tail}
    = \frac{a_\text{tail}}{M} (r^\Delta + r^{N-\Delta}) \sum_{i=0}^\infty r^{iN}
$$

with $r < 1$.
The factor $\frac{1}{M}$ normalizes the tail component so that it sums to $a_\text{tail}$.

$$
M
    &= \left(\frac{1-r^N}{1-r} + r^N \frac{1-r^{-N}}{1-r^{-1}}\right) \sum_{i=0}^\infty r^{iN}
\\
    &= \left(1-r^{-N}\right) \frac{1 + r}{1-r} \sum_{i=0}^\infty r^{iN}
$$

We may absorb the infinite sum into the normalization constant to simplify the model:

$$
c_\Delta^\text{tail}
    & = \frac{a_\text{tail}}{M'} (r^\Delta + r^{N-\Delta})
\\
M'
    &= \left(1-r^{-N}\right) \frac{1 + r}{1-r}
$$

The exponential decay of the tail component is characterized by
its autocorrelation time, $\tau_\text{exp}$ {cite:p}`sokal_1997_monte`:

$$
    r = \exp\left(-\frac{h}{\tau_\text{exp}}\right)
$$

where $h$ is the time step.

### Discrete Fourier Transform

For small values of $k$, the discrete Fourier transform of the model autocorrelation function becomes:

$$
C^\text{exp-tail}_k
    &\approx
    a_\text{short} + \frac{a_\text{tail}}{M'} \left(
        \frac{1-r^N}{1 - r \omega^{-k}} + r^N\frac{1-r^{-N}}{1 - r^{-1} \omega^{-k}}
    \right)
\\
    &\approx
    a_\text{short} + \frac{a_\text{tail}}{M'} \left(
        \frac{1-r^N}{1 - r \omega^{-k}} + \frac{1-r^N}{1 - r \omega^k} - \left(1-r^N\right)
    \right)
\\
    &\approx
    a_\text{short} + \frac{a_\text{tail}}{M'} \left(1-r^N\right) \left(
         \frac{2 - r(\omega^k + \omega^{-k})}{1 - r(\omega^k + \omega^{-k}) + r^2} - 1
    \right)
\\
    &\approx
    a_\text{short} + a_\text{tail} \frac{1-r}{1+r} \left(
         \frac{2 - r(\omega^k + \omega^{-k})}{1 - r(\omega^k + \omega^{-k}) + r^2} - 1
    \right)
$$

For the first term we assumed $\omega^{k\Delta_\text{short}}\approx 1$.
Finally, we can substitute $\omega^k + \omega^{-k} = 2\cos(2\pi k/N)$:

$$
C^\text{exp-tail}_k
    &\approx
    a_\text{short} + a_\text{tail} \frac{1-r}{1+r} \left(
         2\frac{1 - r\cos(2\pi k/N)}{1 - 2r\cos(2\pi k/N) + r^2} - 1
    \right)
    \\
    &\approx
    a_\text{short} + a_\text{tail} \frac{(1-r)^2}{1 - 2r\cos(2\pi k/N) + r^2}
    \\
$$

This model can be fitted to the low-frequency part of an empirical autocorrelation function.
Once the parameters are fitted to an empirical spectrum, one finds:

$$
    \mathcal{I} \approx \hat{\mathcal{I}} = \hat{C}^\text{exp-tail}_0 = \hat{a}_\text{short} + \hat{a}_\text{tail}
$$

### Peak width

As illustrated in the
[Exponential Tail Model](../../examples/model.py)
example, the exponential decay of the autocorrelation function results in
a peak in the power spectrum at zero frequency.
The width of this peak at half the maximum of the tail term is found by solving:

$$
    1 - 2 r \cos(2\pi k_\text{half}/N) + r^2 = 2 (1 - r)^2
$$

The solution is:

$$
    k_\text{half}
        &= \frac{N}{2\pi}\arccos\left(\frac{4r - r^2 -1 }{2r}\right)
    \\
        &= \frac{N}{2\pi}\arccos\left(2 - \cosh\left(\frac{h}{\tau_\text{exp}}\right)\right)
$$

Using RFFT conventions, $f=k/hN$, this can be rewritten as a frequency:

$$
    f_\text{half}
        = \frac{1}{2\pi h}\arccos\left(2 - \cosh\left(\frac{h}{\tau_\text{exp}}\right)\right)
$$

We can rely on the following series expansion:

$$
    \arccos(2 - \cosh(u))
    \approx
    \left|u + \frac{1}{12} u^3 + \frac{1}{96} u^5 + \mathcal{O}(u^7)\right|
$$

Hence, in the limit $\tau_\text{exp} \gg h$, we find:

$$
    f_\text{half}
        \approx \frac{1}{2\pi \tau_\text{exp}}
$$

Sequences obtained from computer simulations
usually have sufficiently small time steps to satisfy this limit:
The steps must be able to resolve the fastest oscillations in the system
to correctly simulate its dynamics.
