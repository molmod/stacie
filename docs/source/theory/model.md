# Model Spectrum

STACIE supports two models for fitting the low-frequency part of the power spectrum.
In both models, the value at zero frequency corresponds to the autocorrelation integral.

1. The [ExpPolyModel](#stacie.model.ExpPolyModel) is the most general;
   it is an exponential function of a linear combination of simple monomials of the frequency.
   You can specify the degrees of the monomials, and typically a low degree works fine:

    - Degree $0$ is suitable for a white noise spectrum.
    - Degrees $\{0, 1\}$ can be used to extract useful information from a noisy spectrum.
    - Degrees $\{0, 1, 2\}$ are applicable to spectra with low statistical uncertainty,
      e.g., averaged over $>100$ inputs.
    - An even polynomial with degrees $\{0, 2\}$ is suitable for spectra
      that are expected to have a vanishing derivative at zero frequency.

    The main advantage of this model is its broad applicability,
    as it requires little prior knowledge of the functional form of the spectrum.

2. The [PadeModel](#stacie.model.PadeModel) is useful in several scenarios:

    - It can be configured to model a spectrum with a Lorentzian peak at the origin
      plus some white noise, which corresponds to an exponentially decaying {term}`ACF`.
      In this case, STACIE also derives the exponential correlation time,
      which can deviate from the integrated correlation time.

    - Rational functions are, in general, interesting because they can be
      parameterized to have well-behaved high-frequency tails,
      which can facilitate the regression.

(section-exppoly-target)=

## 1. ExpPoly Model

The [ExpPolyModel](#stacie.model.ExpPolyModel) is defined as:

$$
    I^\text{exppoly}_k = \exp\left(\sum_{s \in S} b_s f_k^s\right)
$$

where $S$ is the set of polynomial degrees, which must include 0.
With this form, $\exp(b_0)$ corresponds to the integral of the autocorrelation function.
When one obtains an estimate $\hat{b}_0$ and its variance $\hat{\sigma}^2_{b_0}$,
the autocorrelation integral is [log-normally distributed](https://en.wikipedia.org/wiki/Log-normal_distribution)
with estimated mean and variance:

$$
    \begin{aligned}
    \hat{\mathcal{I}}
    &= \exp\left(\hat{b}_0 + \frac{1}{2}\hat{\sigma}^2_{b_0}\right)
    \\
    \hat{\sigma}^2_{\mathcal{I}}
    &= \exp\left(2\hat{b}_0 + \hat{\sigma}^2_{b_0}\right)
        \left(\exp(\hat{\sigma}^2_{b_0}) - 1 \right)
    \end{aligned}
$$

To construct this model, you can create an instance of the `ExpPolyModel` class as follows:

```python
from stacie import ExpPolyModel
model = ExpPolyModel([0, 1, 2])
```

This model is identified as `exppoly(0, 1, 2)` in STACIE's screen output and plots.

(section-pade-target)=

## 2. Pade Model

The [PadeModel](#stacie.model.PadeModel) is defined as:

$$
    I^\text{pade}_k = \frac{
        \displaystyle
        \sum_{s \in S_\text{num}} p_s f_k^s
    }{
        \displaystyle
        1 + \sum_{s \in S_\text{den}} q_s f_k^s
    }
$$

where $S_\text{num}$ contains the polynomial degrees in the numerator, which must include 0,
and $S_\text{den}$ contains the polynomial degrees in the denominator, which must exclude 0.
With this model, $p_0$ corresponds to the integral of the autocorrelation function,
for which we simply have:

$$
    \begin{aligned}
    \hat{\mathcal{I}} &= \hat{p}_0
    \\
    \hat{\sigma}^2_{\mathcal{I}} &= \hat{\sigma}^2_{p_0}
    \end{aligned}
$$

For the special case of a Lorentzian peak at the origin plus some white noise,
that is $S_\text{num} = \{0, 2\}$ and $S_\text{den} = \{2\}$,
the model is equivalent to:

$$
    I^\text{lorentz}_k = A + \frac{B}{1 + (2 \pi f_k \tau_\text{exp})^2}
$$

where $f_k$ is the standard frequency grid of the discrete Fourier transform,
$A$ is the white noise level, $B$ is the amplitude of the Lorentzian peak,
and $\tau_\text{exp}$ is the exponential correlation time.
The frequency grid is defined as $f_k = k / (hN)$,
where $h$ is the time step of the discretized time axis, and $N$ is the number of samples.
We can write the Lorentzian model parameters in terms of the Pade model parameters as follows:

$$
    \begin{aligned}
        A &= \frac{\hat{p}_2}{\hat{q}_2}
        \\
        B &= \hat{p}_0 - \frac{\hat{p}_2}{\hat{q}_2}
        \\
        \tau_\text{exp} &= \frac{\sqrt{q_2}}{2 \pi}
    \end{aligned}
$$

The Pade model will only correspond to a Lorentzian peak if $q_2 > 0$ and $p_0 q_2 > p_2$.
When this is the case, $\tau_\text{exp}$ is related
to the width of the peak ($2 \pi \tau_\text{exp}$) in the power spectrum.
The exponential correlation time and its variance can then be derived
from the fitted parameters with first-order error propagation:

$$
    \begin{aligned}
    \hat{\tau}_\text{exp} &= \frac{\sqrt{\hat{q}_2}}{2 \pi}
    \\
    \hat{\sigma}^2_{\tau_\text{exp}} &= \frac{1}{16 \pi^2 \hat{q}_2} \hat{\sigma}^2_{q_2}
    \end{aligned}
$$

Note that this model is also applicable to data whose short-time correlations are not exponential,
as long as the tail of the ACF decays exponentially.
Such deviating short-time correlations will only affect the white noise level $A$
and features in the PSD at higher frequencies, which will be ignored by STACIE.

To construct this model, you can create an instance of the `PadeModel` class as follows:

```python
from stacie import PadeModel
model = PadeModel([0, 2], [2])
```

This model is identified as `pade(0, 2; 2)` in STACIE's screen output and plots.
