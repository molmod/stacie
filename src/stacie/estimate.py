# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Algorithm to estimate the autocorrelation integral."""

import warnings

import attrs
import numpy as np
from numpy.typing import NDArray
from scipy.optimize import minimize

from .cost import LowFreqCost
from .cutoff import CutoffCriterion, underfitting_criterion
from .model import ExpTailModel, SpectrumModel
from .rpi import build_xgrid_exp, rpi_opt
from .spectrum import Spectrum

__all__ = ("Result", "FCutWarning", "estimate_acint", "fit_model_spectrum")


@attrs.define
class Result:
    """Container class holding all the results of the autocorrelation integral estimate."""

    spectrum: Spectrum = attrs.field()
    """The input spectrum from which the autocorrelation integral is estimated."""

    nfit: int = attrs.field()
    """
    The number of low-frequency spectrum data points used to fit the model.
    This value is the best among a series of tested cutoffs,
    i.e. by minimizing the cutoff criterion.
    """

    history: dict[int, dict[str]] = attrs.field()
    """History of nfit optimization.

    The key is ``nfit``, the number of frequencies fitted to.
    Each value is a dictionary returned by :func:`fit_model_spectrum`.
    """

    @property
    def props(self) -> dict[str]:
        """Properties computed from the fit up to the selected spectrum cutoff.

        This is a shortcut for ``history[nfit]``.
        See return value of :func:`fit_model_spectrum` for more details.
        """
        return self.history[self.nfit]

    @property
    def acint(self) -> float:
        """The autocorrelation integral."""
        return self.props["acint"]

    @property
    def acint_std(self) -> float:
        """The uncertainty of the autocorrelation integral."""
        return self.props["acint_std"]

    @property
    def corrtime_exp(self) -> float:
        """The exponential correlation time."""
        return self.props["corrtime"]

    @property
    def corrtime_exp_std(self) -> float:
        """The uncertainty of the exponential correlation time."""
        return self.props["corrtime_std"]

    @property
    def corrtime_int(self) -> float:
        """The integrated correlation time."""
        return self.props["acint"] / (2 * self.spectrum.prefactor * self.spectrum.variance)

    @property
    def corrtime_int_std(self) -> float:
        """The uncertainty of the integrated correlation time."""
        return self.props["acint_std"] / (2 * self.spectrum.prefactor * self.spectrum.variance)


class FCutWarning(Warning):
    """Raised when there is an issue with the frequency cutoff."""


def estimate_acint(
    spectrum: Spectrum,
    *,
    fcutmax: float | None = None,
    maxscan: int = 100,
    nfitmin: int = 10,
    nfitmax_hard: int = 1000,
    model: SpectrumModel | None = None,
    cutoff_criterion: CutoffCriterion = underfitting_criterion,
    verbose: bool = False,
) -> Result:
    """Estimate the integral of the autocorrelation function.

    It is recommended to leave the keyword arguments to their default values,
    except for methodological testing.

    This function fits a model to the low-frequency portion of the spectrum
    and derives an estimate of the autocorrelation (and its uncertainty) from the fit.
    The model is fitted to a part of the spectrum op to a cutoff frequency.
    Multiple cutoffs are tested, each resulting in another number of spectrum amplitudes in the fit,
    and the one that minimizes the ``cutoff_criterion`` is selected as the best solution.

    Parameters
    ----------
    spectrum
        The power spectrum and related metadata,
        used as inputs for the estimation of the autocorrelation integral.
        This object can be prepared with the function: :py:func:`stacie.spectrum.compute_spectrum`.
    fcutmax
        The maximum cutoff on the frequency axis (units of frequency),
        which corresponds to the largest value for ``nfit``.
    maxscan
        The maximum number of cutoffs to test.
        If 1, then only the given ``fcutmax`` is used.
    nfitmin
        The minimal amount of frequency data points to use in the fit.
    nfitmax_hard
        The maximal amount of frequency data points to use in the fit.
        This puts an upper bound on the computational cost of the fit.
        If this upper limit is stricter than that of ``fcutmax``, a warning is raised.
    model
        The model used to fit the low-frequency part of the spectrum.
        The default is an instance of :py:class:`stacie.model.ExpTailModel`.
    cutoff_criterion
        The criterion function that is minimized to find the best cutoff
        (and thus number of points included in the fit).
    verbose
        Set this to ``True`` to print progress information of the frequency cutoff search
        to the standard output.

    Returns
    -------
    result
        The inputs, intermediate results and outputs or the algorithm.
    """
    if model is None:
        model = ExpTailModel()
    history = {}

    if verbose and maxscan > 1:
        print("   nfit   criterion  incumbent")
        scratch = {}

    def compute_criterion(ifit: int):
        """Criterion to be minimized to find the best frequency cutoff.

        Parameters
        ----------
        ifit
            The index in the nfits list to get the right cutoff.
        include_failed
            When ``True``, add failed fits to the history.
        """
        nfit = nfits[ifit]
        props = fit_model_spectrum(
            spectrum.timestep,
            spectrum.freqs,
            spectrum.amplitudes,
            spectrum.ndofs,
            nfit,
            model,
            cutoff_criterion,
        )
        evals = props["cost_hess_evals"]
        history[nfit] = props
        criterion = (
            props["criterion"] if (np.isfinite(evals).all() and (evals > 0).all()) else np.inf
        )
        if verbose and maxscan > 1:
            lowest_criterion = scratch.get("lowest_criterion")
            best = lowest_criterion is None or criterion < lowest_criterion
            if best:
                scratch["lowest_criterion"] = criterion
            print(f"{nfit:7d}  {criterion:10.1f}  {'<---' if best else ''}")
        return criterion

    nfitmax = len(spectrum.freqs) if fcutmax is None else int(spectrum.freqs.searchsorted(fcutmax))
    if nfitmax > nfitmax_hard:
        nfitmax = nfitmax_hard
        warnings.warn(
            "The maximum frequency cutoff is lowered to constrain "
            f"the maximum number of data points in the fit to {nfitmax}.",
            FCutWarning,
            stacklevel=2,
        )
    if nfitmax < nfitmin:
        raise ValueError("Too few data points for fit.")

    if maxscan == 1:
        nfit = nfitmax
        nfits = [nfit]
        compute_criterion(0)
    else:
        nfits = build_xgrid_exp([nfitmin, nfitmax], maxscan)
        rpi_opt(compute_criterion, [0, len(nfits) - 1], mode="min")
        if any(np.isfinite(props["criterion"]) for props in history.values()):
            nfit = min((record["criterion"], key) for key, record in history.items())[1]
        else:
            warnings.warn(
                "Could not find a suitable frequency cutoff. "
                "The resuts for the smallest cutoff are selected.",
                FCutWarning,
                stacklevel=2,
            )
            nfit = nfits[0]

    return Result(spectrum, nfit, dict(sorted(history.items())))


def fit_model_spectrum(
    timestep: float,
    freqs: NDArray[float],
    amplitudes: NDArray[float],
    ndofs: NDArray[int],
    nfit: int,
    model: SpectrumModel,
    cutoff_criterion: CutoffCriterion,
) -> dict[str, NDArray]:
    """Optimize the parameter of a model for a given spectrum.

    The parameters are the attributes of the :py:class:`stacie.cost.LowFreqCost` class,
    except for the ones documented below.

    Parameters
    ----------
    nfit
        The number of low-frequency data points to use in the fit.
    cutoff_criterion
        The criterion function that is minimized to find the best cutoff
        (and thus number of points included in the fit).

    Returns
    -------
    props
        A dictionary with various intermediate results of the cost function calculation,
        computed for the optimized parameters.
        See Notes for details.

    Notes
    -----
    The returned dictionary contains the following items:

    - ``acint``: the estimate of the autocorrelation integral.
    - ``acint_var``: the variance of the estimate of the autocorrelation integral.
    - ``acint_std``: the standard error of the estimate of the autocorrelation integral.
    - ``corrtime``: the estimate of the slowest time scale in the sequences.
    - ``corrtime_var``: the variance of the estimate of the slowest time scale.
    - ``corrtime_std``: the standard error of the estimate of the slowest time scale.
    - ``cost_value``: the cost function value.
    - ``cost_grad``: the cost Gradient vector (if ``deriv>=1``).
    - ``cost_hess``: the cost Hessian matrix (if ``deriv==2``).
    - ``cost_hess_evals``: the Hessian eigenvalues.
    - ``cost_hess_evecs``: the Hessian eigenvectors.
    - ``covar``: the covariance matrix of the parameters.
    - ``criterion``: the value of the criterion whose minimizer determines the frequency cutoff.
    - ``ll``: the log likelihood.
    - ``pars_init``: the initial guess of the parameters.
    - ``pars``: the optimized parameters.
    - ``thetas``: scale parameters for the gamma distribution.
    - ``timestep``: the time step.
    """
    # Maximize likelihood
    model.precondition(timestep, amplitudes[:nfit])
    pars_init = model.guess(freqs[:nfit], amplitudes[:nfit], ndofs[:nfit])
    if not model.valid(pars_init):
        raise AssertionError("Infeasible guess")
    cost = LowFreqCost(timestep, freqs[:nfit], amplitudes[:nfit], ndofs[:nfit], model)
    opt = minimize(
        cost.funcgrad,
        pars_init,
        jac=True,
        hess=cost.hess,
        bounds=model.bounds(),
        method="trust-constr",
        options={"xtol": 1e-10, "gtol": 1e-10},
    )
    props = cost.props(opt.x, 2)

    # Compute the Hessian and its properties.
    evals, evecs = np.linalg.eigh(props["cost_hess"])
    props["cost_hess_evals"] = evals
    props["cost_hess_evecs"] = evecs
    if (evals > 0).all() and np.isfinite(evals).all():
        half = evecs / np.sqrt(evals)
        props["covar"] = np.dot(half, half.T)
    else:
        props["covar"] = np.full_like(props["cost_hess"], np.inf)

    # Derive estimates from model parameters.
    props.update(model.derive_props(props["pars"], props["covar"]))

    # Compute remaining properties and derive the cutoff criterion
    props["pars_init"] = pars_init
    props["freqs_rest"] = freqs[nfit:]
    props["amplitudes_rest"] = amplitudes[nfit:]
    props["kappas_rest"] = 0.5 * ndofs[nfit:]
    props["criterion"] = cutoff_criterion(props)

    # Remove some intermediate properties to reduce the size of the Result object.
    del props["freqs"]
    del props["kappas"]
    del props["amplitudes"]
    del props["freqs_rest"]
    del props["amplitudes_rest"]
    del props["kappas_rest"]

    return props
