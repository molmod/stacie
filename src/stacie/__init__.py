# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024-2025 The contributors of the Stacie Python Package.
# See the CONTRIBUTORS.md file in the project root for a full list of contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""The Stacie package."""

from __future__ import annotations

from .estimate import estimate_acint
from .plot import UnitConfig, plot_results
from .spectrum import compute_spectrum
from .summary import summarize_results

__all__ = (
    "UnitConfig",
    "__version__",
    "__version_tuple__",
    "compute_spectrum",
    "estimate_acint",
    "plot_results",
    "summarize_results",
)


try:
    from ._version import __version__, __version_tuple__
except ImportError:
    __version__ = "0.0.0a-dev"
    __version_tuple__ = (0, 0, 0, "a-dev")
