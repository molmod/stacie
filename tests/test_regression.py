# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Regression tests for typical Stacie workflows."""

from contextlib import ExitStack

import pytest
from path import Path
from stacie.estimate import FCutWarning, estimate_acfint
from stacie.model import ExpTailModel, SpectrumModel, WhiteNoiseModel
from stacie.msgpack import dump, load
from stacie.plot import plot
from stacie.spectrum import Spectrum

DOUBLE_NAMES = ["double1", "double2"]
WHITE_NAMES = ["white1", "white2"]
NAME_LISTS = DOUBLE_NAMES, WHITE_NAMES
ALL_NAMES = [j for i in NAME_LISTS for j in i]


def output_test_result(prefix, res):
    dn_out = Path("tests/outputs")
    dn_out.makedirs_p()
    path_pdf = dn_out / f"{prefix}.pdf"
    plot(path_pdf, res)
    path_zip = dn_out / f"{prefix}.nmpk.xz"
    dump(path_zip, res)


def register_result(regtest, res):
    with regtest:
        print(f"ACF Int  = {res.props['acfint']:.4e} ± {res.props['acfint_std']:.3e}")
        print(f"Tau tail = {res.props['corrtime_tail']:.4e} ± {res.props['corrtime_tail_std']:.3e}")
        print(f"Log lh   = {res.props['ll']:.5e}")
        print("---")


def check_noscan_single(
    regtest,
    spectrum: Spectrum,
    prefix: str,
    fcutmax: float = 0.005,
    model: SpectrumModel | None = None,
):
    res = estimate_acfint(spectrum, fcutmax=fcutmax, maxscan=1, model=model)
    register_result(regtest, res)
    output_test_result(prefix, res)


@pytest.mark.parametrize("name", ALL_NAMES)
def test_epxtail_noscan(regtest, name):
    spectrum = load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum)
    check_noscan_single(regtest, spectrum, f"epxtail_noscan_{name}")


@pytest.mark.parametrize("name", WHITE_NAMES)
def test_white_noscan(regtest, name):
    spectrum = load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum)
    check_noscan_single(
        regtest, spectrum, f"white_noscan_{name}", fcutmax=0.1, model=WhiteNoiseModel()
    )


@pytest.mark.parametrize(("name", "fcutmax"), [("white2", 0.008), ("double1", 0.05)])
def test_exptail_noscan_fail(regtest, name, fcutmax):
    spectrum = load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum)
    check_noscan_single(regtest, spectrum, f"exptail_noscan_{name}_fail", fcutmax=fcutmax)


@pytest.mark.parametrize("names", NAME_LISTS)
def test_exptail_noscan_multi(regtest, names):
    res = []
    for name in names:
        spectrum = load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum)
        r = estimate_acfint(spectrum, fcutmax=0.005, maxscan=1)
        register_result(regtest, r)
        res.append(r)
    output_test_result("exptail_noscan_multi", res)


@pytest.mark.parametrize("name", ALL_NAMES)
def test_exptail_scan(regtest, name):
    spectrum = load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum)
    res = estimate_acfint(spectrum, fcutmax=0.03, maxscan=10)
    register_result(regtest, res)
    output_test_result(f"exptail_scan_{name}", res)


@pytest.mark.parametrize("names", NAME_LISTS)
@pytest.mark.parametrize("model", [ExpTailModel(), WhiteNoiseModel()])
@pytest.mark.parametrize("zero_freq", [False, True])
def test_scan_multi(regtest, zero_freq, model, names):
    should_warn = names[0].startswith("double") and isinstance(model, WhiteNoiseModel)
    res = []
    for name in names:
        spectrum = load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum)
        if not zero_freq:
            spectrum = spectrum.without_zero_freq()
        with ExitStack() as stack:
            if should_warn:
                stack.enter_context(pytest.warns(FCutWarning))
            r = estimate_acfint(spectrum, fcutmax=0.01, maxscan=10, model=model)
        register_result(regtest, r)
        res.append(r)
    prefix = f"{model.name}_scan_multi"
    if not zero_freq:
        prefix += "_nodc"
    output_test_result(prefix, res)
