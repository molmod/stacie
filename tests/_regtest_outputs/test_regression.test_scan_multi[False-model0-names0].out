ACF Int  = 1.4142e+02 ± 5.996e+00
Tau tail = 2.0411e+02 ± 2.207e+01
Log lh   = -3.83948e+01
---
ACF Int  = 1.4422e+02 ± 6.624e+00
Tau tail = 2.7431e+02 ± 3.428e+01
Log lh   = -4.64089e+01
---
